### PR TITLE
feat(tree): add tree demo page

### PR DESCRIPTION
**Pull request by tinayuangao**

### CODE REVIEW

File: src/demo-app/tree/tree-demo.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    this.nestedDataSource = new JsonNestedDataSource(database);
+  }
+
+  getLevel = (node: JsonFlatNode) => { return node.level };
+
+  isExpandable = (node: JsonFlatNode) => { return node.expandable; }
+
+  getChildren = (node: JsonNode) => { return ofObservable(node.children); }
+
+  hasChild = (_: number, _nodeData: JsonFlatNode) => { return _nodeData.expandable; }
```
Comment (benb7760) [inline src/demo-app/tree/tree-demo.ts]: @tinayuangao I noticed this a week or so ago while trialling the tree; users have to define two functions that achieve the exact same result but with different signatures; `isExpandable`, used by the `when`, and `hasChild`, used by the TreeControl. Would it be sensible to change this?

File: src/demo-app/tree/tree-demo.html [hunk 2]
```diff
@@ -0,0 +1,47 @@
+<div class="demo-tree-container">
+  <mat-card>
+    <mat-card-header>Flattened tree</mat-card-header>
+
+    <mat-tree [dataSource]="dataSource" [treeControl]="treeControl">-->
```
Comment (jelbourn) [inline src/demo-app/tree/tree-demo.html]: Hanging `-->` ?

File: src/demo-app/tree/tree-demo.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  // Flat tree data source
+  dataSource: FlatDataSource;
+
+  // Nested tree data source
+  nestedDataSource: JsonNestedDataSource;
+
+  constructor(database: JsonDatabase) {
+    // For flat tree
+    this.treeControl = new FlatTreeControl<JsonFlatNode>(this.getLevel, this.isExpandable);
```
Comment (jelbourn) [inline src/demo-app/tree/tree-demo.ts]: Would it make sense to use a more real-world example, such as collection of directories?

**Comment by tinayuangao** [comment]
Comments addressed. Please take a look. Thanks!

**Pull request merged**